src/modules/schnorrsig/main_impl.h

```diff
@@ -264,4 +264,291 @@ int secp256k1_schnorrsig_verify(const secp256k1_context* ctx, const unsigned cha
+        return 0;
+    }
+
+    /* Write common prefix pre into tagged hash  */
+    /* Let n' = starting_from */
+    /* pre = (r_0 || pk_0 || mes_0 || .... ||  r_{n'-1} || pk_{n'-1} || mes_{n'-1}) */
+    secp256k1_schnorrsig_sha256_tagged_aggregation(&hash);
+    for (i = 0; i < starting_from; ++i) {
+        /* write r_i */
+        secp256k1_sha256_write(&hash, rs[i], 32);
+        /* write pk_i */
+        secp256k1_sha256_write(&hash, &pubkeys_ser[i*32], 32);
+        /* write mes_i*/
+        secp256k1_sha256_write(&hash, &msgs32[i*32], 32);
+    }
+
+    /* zi = Hash( pre || r_{n'} || pk_{n'} || mes_{n'} || ... || r_{i} || pk_{i} || mes_{i} ) */
+    for (i = starting_from; i < n_sigs; ++i) {
+        unsigned char hashoutput[32];
+        secp256k1_sha256 hashcopy; 
+        /* Write into hash r_i, pk_i, mes_i, r_i */
+        secp256k1_sha256_write(&hash, rs[i], 32);
+        secp256k1_sha256_write(&hash, &pubkeys_ser[i*32], 32);
+        secp256k1_sha256_write(&hash, &msgs32[i*32], 32);
+        /* Copy the hash */
+        memcpy(&hashcopy,&hash,sizeof(hash));
+        /* Finalize the copy to get zi*/
+        secp256k1_sha256_finalize(&hashcopy, hashoutput);
+        secp256k1_scalar_set_b32(&zs[i-starting_from], hashoutput, NULL);
+    }
+    return 1;
+}
+
+
+int secp256k1_schnorrsig_inc_aggregate(const secp256k1_context* ctx, unsigned char* aggsig, size_t* aggsig_size, const secp256k1_xonly_pubkey* all_pubkeys, const unsigned char* all_msgs32, const unsigned char* new_sigs64, size_t n_sigs_before, size_t n_sigs_new) {
+    uint32_t i;
+    uint32_t n_sigs;
+    int overflow;
+    secp256k1_scalar s;
+    unsigned char* pubkeys_ser; /* this will store serialized pubkeys */
+    secp256k1_scalar* zs; /* this will store the aggregation coefficients */ 
+    const unsigned char** rs; /* this will store pointers to the r's in the aggsig */
+
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(aggsig != NULL);
+    ARG_CHECK(aggsig_size != NULL);
+    ARG_CHECK(all_pubkeys != NULL);
+    ARG_CHECK(all_msgs32 != NULL);
+    ARG_CHECK(new_sigs64 != NULL);
+
+    /* Check that aggsig_size is large enough, i.e. aggsig_size >= 32*(n_sigs+1) */
+    n_sigs = n_sigs_before + n_sigs_new;
+    if ((*aggsig_size >> 5) <= 0 || ((*aggsig_size >> 5)-1) < n_sigs) {
+        return 0;
+    }
+
+    /* Serialize public keys. */ 
+    /* We need that for the aggregation coefficients */
+    pubkeys_ser = (unsigned char*) checked_malloc(&ctx->error_callback, 32*n_sigs); 
+    for (i = 0; i < n_sigs; ++i) {
+        if (!secp256k1_xonly_pubkey_serialize(ctx, &pubkeys_ser[i*32], &all_pubkeys[i])) {
+            free(pubkeys_ser);
+            return 0;
+        }
+    }
+
+    /* Compute new aggregation coefficients zi*/
+    zs = (secp256k1_scalar*) checked_malloc(&ctx->error_callback, n_sigs_new*sizeof(secp256k1_scalar));
+    rs = (const unsigned char**) checked_malloc(&ctx->error_callback, n_sigs*sizeof(unsigned char*));
```

Comment:
Oh, hm, so it's a policy (like in secp256k1) that we don't call malloc/free (except for context creation, but this can also be done in a block of memory allocated by the caller, see the `secp256k1_prealloc.h` header). The reasons are safety and support for embedded systems.

I haven't thought this through entirely, but I think all algorithms in the spec can be implemented without dynamic allocation (i.e., in O(1) stack memory), by rearranging the loops. 

At least this is true as long as we don't multi-multiplications. For multi-multiplications, we'll necessarily need some form of memory management. In upstream secp256k1, we have a special data type, but it's unused there, and we may change this in the future, because we're not particularly happy with it. See https://github.com/bitcoin-core/secp256k1/issues/1302.

So I tend to suggest rewriting the loops (if possible) and ignoring multi-multiplications for now.  @jonasnick What do you think?